rust/template/distributed_datalog/observe/src/observable.rs

```diff
@@ -34,17 +43,80 @@ where
     T: Debug + Send + 'static,
     E: Debug + Send + 'static,
 {
-    /// An observer subscribes to the delta stream from an outlet.
-    fn subscribe(&mut self, observer: ObserverBox<T, E>) -> Option<Box<dyn Subscription>> {
+    type Subscription = ();
+
+    fn subscribe(
+        &mut self,
+        observer: ObserverBox<T, E>,
+    ) -> Result<Self::Subscription, ObserverBox<T, E>> {
         let mut guard = self.observer.lock().unwrap();
         match *guard {
-            Some(_) => None,
+            Some(_) => Err(observer),
             None => {
                 *guard = Some(observer);
-                Some(Box::new(UpdatesSubscription {
-                    observer: self.observer.clone(),
-                }))
+                Ok(())
             }
         }
     }
+
+    fn unsubscribe(&mut self, _subscription: &Self::Subscription) -> Option<ObserverBox<T, E>> {
+        let mut guard = self.observer.lock().unwrap();
+        match *guard {
```

Comment:
Can you use `swap(guard, &None)` or something like that instead of the `match`?

Reply:
Thanks for taking a look! As part of the changes I have piled up (which will go out later today or on Monday) the `match` will go and the `unsubscribe` will actually become a one-liner:
```rust
self.observer.lock().unwrap().take()
```
So I'd like to just leave that be right now, as it's merely a transitional state.